src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,115 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double CAPACITY_MULTIPLY_BY = 1.5;
+    private static final String NEGATIVE_INDEX_ERROR = "Index should be natural number";
+    private static final String INDEX_OUT_OF_SIZE_ERROR = "Index out of list length";
+
+    private T[] data;
+    private int size;
+
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity > 0) {
+            data = (T[]) new Object[capacity];
+        } else if (capacity == 0) {
+            data = (T[]) new Object[]{};
+        } else {
+            throw new IllegalArgumentException(NEGATIVE_INDEX_ERROR);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= data.length) {
+            increaseCapacity();
+        }
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        isIndexValid(index);
+        if (size >= data.length) {
+            increaseCapacity();
+        }
+        System.arraycopy(data, index, data, index + 1, size - index);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index + 1);
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index + 1);
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index + 1);
+        final T result = data[index];
+        System.arraycopy(data, index + 1, data, index, size - index - 1);
+        data[size - 1] = null;
+        size--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if ((data[i] != null && data[i].equals(element))
+                    || (data[i] == null && element == null)) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException(INDEX_OUT_OF_SIZE_ERROR);
+        }
```

Comment:
is this out of bounds or, maybe, you should show a message, that element doesn't exist? Don't forget to make your exception message informative